File: artifacts/pagespeed-reupload/upload-via-admin.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';
import { chromium } from 'playwright';

const projectRoot = path.resolve('C:/Users/Sadjad Rgz/Documents/Majlesyar2');
const manifestPath = path.join(projectRoot, 'artifacts', 'pagespeed-reupload', 'manifest.json');
const screenshotDir = path.join(projectRoot, 'artifacts', 'pagespeed-reupload', 'screenshots');
const baseUrl = 'https://majlesyar.com';
const adminBase = `${baseUrl}/majmanage`;
const username = process.env.MAJLESYAR_ADMIN_USER || 'admin';
const password = process.env.MAJLESYAR_ADMIN_PASSWORD || 'admin';

async function readManifest() {
  const raw = await fs.readFile(manifestPath, 'utf8');
  return JSON.parse(raw);
}

async function ensureDir(dir) {
  await fs.mkdir(dir, { recursive: true });
}

async function login(page) {
  await page.goto(`${adminBase}/login/?next=/majmanage/`, { waitUntil: 'domcontentloaded' });
  await page.locator('input[name="username"]').fill(username);
  await page.locator('input[name="password"]').fill(password);
  await Promise.all([
    page.waitForURL(/\/majmanage\/$/),
    page.locator('button[type="submit"], input[type="submit"]').click(),
  ]);
}

async function verifyProductApi(item) {
  const response = await fetch(`${baseUrl}/api/v1/products/`, {
    headers: { accept: 'application/json' },
  });
  if (!response.ok) {
    throw new Error(`API check failed for ${item.slug}: ${response.status}`);
  }

  const products = await response.json();
  const current = products.find((entry) => entry.id === item.id || entry.url_slug === item.slug);
  if (!current) {
    throw new Error(`Could not find ${item.slug} in API response`);
  }

  if (current.image_name !== item.image_name) {
    throw new Error(`image_name changed for ${item.slug}: ${current.image_name}`);
  }
  if (current.image_alt !== item.image_alt) {
    throw new Error(`image_alt changed for ${item.slug}: ${current.image_alt}`);
  }
  if (!String(current.image || '').toLowerCase().endsWith('.webp')) {
    throw new Error(`image URL for ${item.slug} is not webp: ${current.image}`);
  }

  const imageResponse = await fetch(current.image, { redirect: 'follow' });
  const contentType = imageResponse.headers.get('content-type') || '';
  if (!imageResponse.ok) {
    throw new Error(`image URL failed for ${item.slug}: ${imageResponse.status}`);
  }
  if (!contentType.toLowerCase().includes('image/webp')) {
    throw new Error(`unexpected content type for ${item.slug}: ${contentType}`);
  }

  return {
    image: current.image,
    contentType,
  };
}

async function uploadProduct(page, item) {
  const changeUrl = `${adminBase}/catalog/product/${item.id}/change/`;
  const uploadPath = path.join(projectRoot, item.upload_ready);

  await page.goto(changeUrl, { waitUntil: 'domcontentloaded' });
  await page.locator('#id_name').waitFor();

  const [currentName, currentImageName, currentAlt] = await Promise.all([
    page.locator('#id_name').inputValue(),
    page.locator('#id_image_name').inputValue(),
    page.locator('#id_image_alt').inputValue(),
  ]);

  if (currentName !== item.name) {
    throw new Error(`Unexpected name on admin form for ${item.slug}: ${currentName}`);
  }
  if (currentImageName !== item.image_name) {
    throw new Error(`Unexpected image_name on admin form for ${item.slug}: ${currentImageName}`);
  }
  if (currentAlt !== item.image_alt) {
    throw new Error(`Unexpected image_alt on admin form for ${item.slug}: ${currentAlt}`);
  }

  await page.locator('#id_image').setInputFiles(uploadPath);

  await Promise.all([
    page.waitForLoadState('networkidle'),
    page.locator('[name="_save"]').click(),
  ]);

  const successMessage = page.locator('.messagelist .success, ul.messagelist li.success');
  await successMessage.waitFor();
}

async function screenshotPublicProduct(context, item) {
  const page = await context.newPage();
  await page.goto(`${baseUrl}/product/${item.slug}`, { waitUntil: 'networkidle' });
  await page.screenshot({
    path: path.join(screenshotDir, `${item.slug}.png`),
    fullPage: true,
  });
  await page.close();
}

async function main() {
  const manifest = await readManifest();
  await ensureDir(screenshotDir);

  const browser = await chromium.launch({
    channel: 'chrome',
    headless: true,
  });

  try {
    const adminContext = await browser.newContext();
    const adminPage = await adminContext.newPage();
    await login(adminPage);

    for (const item of manifest) {
      console.log(`Uploading ${item.slug} using ${item.upload_filename}`);
      await uploadProduct(adminPage, item);
      const verification = await verifyProductApi(item);
      console.log(`Verified ${item.slug}: ${verification.image} (${verification.contentType})`);
    }

    const publicContext = await browser.newContext();
    for (const item of manifest) {
      await screenshotPublicProduct(publicContext, item);
    }
    await publicContext.close();
    await adminContext.close();
  } finally {
    await browser.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: artifacts/pagespeed-reupload/upload-via-admin.py
from __future__ import annotations

import json
import os
import time
from pathlib import Path

import requests
from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
from playwright.sync_api import sync_playwright


PROJECT_ROOT = Path(r"C:\Users\Sadjad Rgz\Documents\Majlesyar2")
REUPLOAD_ROOT = Path(
    os.environ.get(
        "MAJLESYAR_REUPLOAD_ROOT",
        str(PROJECT_ROOT / "artifacts" / "pagespeed-reupload"),
    )
)
MANIFEST_PATH = REUPLOAD_ROOT / "manifest.json"
SCREENSHOT_DIR = REUPLOAD_ROOT / "screenshots"
BASE_URL = "https://majlesyar.com"
ADMIN_BASE = f"{BASE_URL}/majmanage"
USERNAME = os.environ.get("MAJLESYAR_ADMIN_USER", "admin")
PASSWORD = os.environ.get("MAJLESYAR_ADMIN_PASSWORD", "admin")
FORCE_UPLOAD_SLUGS = {
    slug.strip()
    for slug in os.environ.get("MAJLESYAR_FORCE_UPLOAD_SLUGS", "").split(",")
    if slug.strip()
}


def read_manifest() -> list[dict]:
    return json.loads(MANIFEST_PATH.read_text(encoding="utf-8"))


def fetch_product_snapshot(item: dict) -> dict:
    response = requests.get(f"{BASE_URL}/api/v1/products/", timeout=30)
    response.raise_for_status()
    products = response.json()
    current = next(
        (entry for entry in products if entry.get("id") == item["id"] or entry.get("url_slug") == item["slug"]),
        None,
    )
    if not current:
        raise RuntimeError(f"Could not find {item['slug']} in API response")
    return current


def verify_product_api(item: dict, snapshot: dict) -> dict:
    current = fetch_product_snapshot(item)

    if current.get("name") != snapshot.get("name"):
        raise RuntimeError(f"name changed for {item['slug']}: {current.get('name')}")
    if current.get("image_name") != snapshot.get("image_name"):
        raise RuntimeError(f"image_name changed for {item['slug']}: {current.get('image_name')}")
    if current.get("image_alt") != snapshot.get("image_alt"):
        raise RuntimeError(f"image_alt changed for {item['slug']}: {current.get('image_alt')}")

    image_url = str(current.get("image") or "")
    if not image_url.lower().endswith(".webp"):
        raise RuntimeError(f"image URL for {item['slug']} is not webp: {image_url}")

    image_response = requests.get(image_url, timeout=30)
    image_response.raise_for_status()
    content_type = image_response.headers.get("content-type", "")
    if "image/webp" not in content_type.lower():
        raise RuntimeError(f"unexpected content type for {item['slug']}: {content_type}")

    return {"image": image_url, "content_type": content_type}


def upload_product(page, item: dict, snapshot: dict) -> None:
    change_url = f"{ADMIN_BASE}/catalog/product/{item['id']}/change/"
    upload_path = PROJECT_ROOT / item["upload_ready"]

    page.goto(change_url, wait_until="domcontentloaded")
    page.locator("#id_name").wait_for()

    current_name = page.locator("#id_name").input_value()
    current_image_name = page.locator("#id_image_name").input_value()
    current_alt = page.locator("#id_image_alt").input_value()

    if current_name != snapshot.get("name"):
        raise RuntimeError(f"Unexpected name on admin form for {item['slug']}: {current_name}")
    if current_image_name != snapshot.get("image_name"):
        raise RuntimeError(f"Unexpected image_name on admin form for {item['slug']}: {current_image_name}")
    if current_alt != snapshot.get("image_alt"):
        raise RuntimeError(f"Unexpected image_alt on admin form for {item['slug']}: {current_alt}")

    page.locator("#id_image").set_input_files(
        [
            {
                "name": item["upload_filename"],
                "mimeType": "image/webp",
                "buffer": upload_path.read_bytes(),
            }
        ]
    )
    page.locator('[name="_save"]').click()
    page.wait_for_timeout(4000)

    if page.locator(".errornote, ul.errorlist").count():
        raise RuntimeError(f"Validation error shown after saving {item['slug']}")


def screenshot_public_product(context, item: dict) -> None:
    page = context.new_page()
    page.goto(f"{BASE_URL}/product/{item['slug']}", wait_until="domcontentloaded", timeout=60000)
    page.wait_for_timeout(5000)
    page.screenshot(path=str(SCREENSHOT_DIR / f"{item['slug']}.png"), full_page=True)
    page.close()


def main() -> None:
    manifest = read_manifest()
    SCREENSHOT_DIR.mkdir(parents=True, exist_ok=True)

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(channel="chrome", headless=True)
        admin_context = browser.new_context()
        login_page = admin_context.new_page()
        login_page.goto(f"{ADMIN_BASE}/login/?next=/majmanage/", wait_until="domcontentloaded")
        login_page.locator('input[name="username"]').fill(USERNAME)
        login_page.locator('input[name="password"]').fill(PASSWORD)
        login_page.locator('button[type="submit"], input[type="submit"]').click()
        login_page.wait_for_url("**/majmanage/")
        login_page.close()

        for item in manifest:
            snapshot = fetch_product_snapshot(item)
            current_image = str(snapshot.get("image") or "").lower()
            should_force_upload = item["slug"] in FORCE_UPLOAD_SLUGS
            if current_image.endswith(".webp") and not should_force_upload:
                print(f"Skipping upload for {item['slug']} because it is already webp")
            elif current_image.endswith((".jpg", ".jpeg")) or should_force_upload:
                print(f"Uploading {item['slug']} using {item['upload_filename']}")
                page = admin_context.new_page()
                try:
                    upload_product(page, item, snapshot)
                finally:
                    page.close()
            else:
                raise RuntimeError(f"Unexpected source image format for {item['slug']}: {snapshot.get('image')}")

            verification = None
            last_error = None
            for _ in range(12):
                try:
                    verification = verify_product_api(item, snapshot)
                    break
                except Exception as exc:  # noqa: BLE001
                    last_error = exc
                    time.sleep(5)
            if verification is None:
                raise RuntimeError(f"Verification failed for {item['slug']}: {last_error}") from last_error
            print(f"Verified {item['slug']}: {verification['image']} ({verification['content_type']})")

        public_context = browser.new_context()
        for item in manifest:
            screenshot_public_product(public_context, item)
        public_context.close()

        admin_context.close()
        browser.close()


if __name__ == "__main__":
    main()
